_discussion_reason_codes.md

```diff
@@ -0,0 +1,108 @@
+| Python     | [ServiceBusError(base)][python_exception] | yes, via the type of the exception itself |                                                          |
+| Java       | [ServiceBusException][java_exception]     | no                                        |                                                          |
+
+## Additional questions
+
+- [ ] Are users expected to react to these errors? If so, what other flags/information might we need?
+- [ ] Is Java going with a base-class so users distinguish via types (like Python) or with an enum?
+- [ ] Here's the current list of Reasons and their descriptions. Do we like these? Can we sync names in our code?
+
+```csharp
+  public enum ServiceBusFailureReason
+    {
+        /// <summary>
+        /// The exception was the result of a general error within the client library.
+        /// </summary>
+        GeneralError,
+
+        /// <summary>
+        /// A Service Bus resource cannot be found by the Service Bus service.
+        /// </summary>
+        MessagingEntityNotFound,
+
+        /// <summary>
+        /// The lock on the message is lost. Callers should call attempt to receive and process the message again.
+        /// </summary>
+        MessageLockLost,
+
+        /// <summary>
+        /// The requested message was not found.
+        /// </summary>
+        MessageNotFound,
+
+        /// <summary>
+        /// A message is larger than the maximum size allowed for its transport.
+        /// </summary>
+        MessageSizeExceeded,
+
+        /// <summary>
+        /// The Messaging Entity is disabled. Enable the entity again using Portal.
+        /// </summary>
+        MessagingEntityDisabled,
+
+        /// <summary>
+        /// The quota applied to an Service Bus resource has been exceeded while interacting with the Azure Service Bus service.
+        /// </summary>
+        QuotaExceeded,
+
+        /// <summary>
+        /// The Azure Service Bus service reports that it is busy in response to a client request to perform an operation.
+        /// </summary>
+        ServiceBusy,
+
+        /// <summary>
+        /// An operation or other request timed out while interacting with the Azure Service Bus service.
+        /// </summary>
+        ServiceTimeout,
+
+        /// <summary>
+        /// There was a general communications error encountered when interacting with the Azure Service Bus service.
+        /// </summary>
+        ServiceCommunicationProblem,
+
+        /// <summary>
+        /// The requested session cannot be locked.
+        /// </summary>
+        SessionCannotBeLocked,
+
+        /// <summary>
+        /// The lock on the session has expired. Callers should request the session again.
+        /// </summary>
+        SessionLockLost,
+
+        /// <summary>
+        /// The user doesn't have access to the entity.
+        /// </summary>
+        Unauthorized,
```

Comment:
This may be worthy of discussion.   Feedback during Event Hubs was that we should use the natural exception in the language, if one applies.    In this case, we were asked to use an `UnauthorizedAccessException`. which is also the behavior from the EH and SB clients in Track One.

Since we're using language types like `ArgumentException`, `ArgumentOutOfRangeException`, `InvalidOperationException`, and `NotSupportedException` rather than wrapping them in `ServiceBusException`,  is there a reason that we don't want to do the same for the unauthorized scenario?

Reply:
Track 1 uses a custom UnathorizedException that derives from ServiceBusException.
I think the distinction of when we use ServiceBusException is when the exception actually comes from the service.

Reply:
Except for the other ones that I mentioned, which are in the AMQP mapping but go to standard language types.  :smile:

Reply:
Fair point.

Reply:
Personally, I don't feel strongly.    Krzysztof did, at one point any may have guidance here.

Reply:
Yeah I think we should probably be consistent here for .NET.

Reply:
@KieranBrantnerMagee , I'm guessing that Python might have a similar concern but I don't know if there's a stdlib type exception that would cover this case (so you would have to make your own anyways).

Reply:
@JoshLove-msft  - and for my own curiosity - are you saying you're going to swap over to using your native language exception here? I believe this change would only affect .net. @hemanttanwar  - does Java have a similar built-in exception for this as well?

Reply:
Correct - there is no reason to differ from EH here, especially when we are already using native language exceptions for some service errors. If we were going to be consistent and always use SBException/EHException for service errors, it would be a different story.